Limit number of objects on topology views

zeari opened the pull request:
Set object limit in settings:
![screencapture-localhost-3000-configuration-index-1485269167360](https://cloud.githubusercontent.com/assets/11256940/22251744/8f4f145c-e254-11e6-8350-b1a8c88685c5.png)

Objects kinds are hidden by order of granularity. Since `containers` are the least major entity they will be hidden first, then pods, then similarly other types until we are below the set object limit.
Edit: you can bring back each hidden type separately (containers, pods, etc.) using the filter buttons on top
![screencapture-localhost-3000-container_topology-show-1483886321675](https://cloud.githubusercontent.com/assets/11256940/21750643/dd032904-d5c0-11e6-913d-22bfeba52938.png)


@dclarizio @simon3z The code here is bad but in general, this is a possible approach towards scaling in topology. If this policy is good enough Ill go forward with it.

zeari commented:
@miq-bot add_label providers/containers, topology

zeari commented:
Setting the default number of items per page for topology to 20 would yield:
![screencapture-localhost-3000-container_topology-show-1483886921445](https://cloud.githubusercontent.com/assets/11256940/21750715/43917f6c-d5c2-11e6-9144-abfaeeec1197.png)

simon3z commented:
Approach LGTM :+1:. 

@dclarizio can you review the approach/screenshots? (Code can be improved when we'll move this out of WIP).

@miq-bot assign dclarizio

dclarizio commented:
@simon3z I'm not sure I'm fully on board with the approach.  I mean, in theory I like it, but I have some issues:
* you can't tell if there are more nodes you are not seeing, so it could be misleading, especially if you see only some of a certain node type
* if I'm at my limit and I try to turn on another node type, nothing will happen, so we may need to show an error message

When I was thinking about this earlier, I had an idea that we could limit the maximum number of child nodes of any type (per parent), then show an extra "group" node, something like `28 more...`.

Another limit setting, at least for the initial display, could be `Child Node Depth`, which would turn off any nodes more than X levels below the root.

In both cases, perhaps we could let the user then drill into (or turn on) missing nodes on demand if they really need to see them.

Adding @serenamarie125 for more input.

zeari commented:
> you can't tell if there are more nodes you are not seeing, so it could be misleading, especially if you see only some of a certain node type
> if I'm at my limit and I try to turn on another node type, nothing will happen, so we may need to show an error message

@dclarizio This is not the case. In this solution, all of a certain type are hidden (meaning all containers\ all pods\ all of type) by default and you can bring them all back using the filtering buttons on top. Those filter buttons are just on by default. Its indicated by the greyed out icons and the user can return to have all types shown. 

Node depth from provider might not be a better filtering method since there are many types that are 'the same depth' but are less granular or interesting than their counterparts.

dclarizio commented:
@zeari ahhh, I see . .. I misunderstood . . . so we eliminate (turn off) types, based on the furthest out nodes (i.e. granularity) until below the configured limit, but the user can turn them back on if they want, cool.

I still think a max number of child nodes would be a good idea, such that if I know more than 50 nodes is impossible for me to see clearly, it would just show the 50 and then a `83 more` node that I could drill into if I wanted to.  Just an idea. 😄

zeari commented:
> I still think a max number of child nodes would be a good idea, such that if I know more than 50 nodes is impossible for me to see clearly, it would just show the 50 and then a 83 more node that I could drill into if I wanted to. Just an idea.

That might be nice. Im not sure how we would decide which entity to hide (node?container?pod?) or how many of them to hide.
Anyway, if this approach is good enough for now I can go forward with it @dclarizio

dclarizio commented:
@zeari this will help and can potentially be improved on later, I'm on board with moving forward . . . Thx, Dan

serenamarie125 commented:
@zeari Couple of questions ...

1. If they limit the number of objects to see by default, is there to way to bring some back into view incrementally without leaving the page?

2. Does it make sense to limit the number where you are ?  It looks like it's the same place as setting number of items per page (as in paging)

zeari commented:
> If they limit the number of objects to see by default, is there to way to bring some back into view incrementally without leaving the page?

@serenamarie125 yes, you can bring back each hidden type separately (`containers`, `pods`, etc.) using the filter buttons on top

> Does it make sense to limit the number where you are ? It looks like it's the same place as setting number of items per page (as in paging)

Maybe not. Where would you put it?

himdel commented on app/assets/javascripts/controllers/container_topology/container_topology_controller.js:
```diff
@@ -37,6 +37,20 @@ function ContainerTopologyCtrl($scope, $http, $interval, $location, topologyServ
       if (currentSelectedKinds && (Object.keys(currentSelectedKinds).length != Object.keys($scope.kinds).length)) {
         $scope.kinds = currentSelectedKinds;
+      } else {
+        const remove_hierarchy = ['Container', 'ContainerGroup', 'ContainerReplicator', 'ContainerService',
```
can't use `const` yet, sorry :)

himdel commented on app/assets/javascripts/controllers/container_topology/container_topology_controller.js:
```diff
@@ -37,6 +37,20 @@ function ContainerTopologyCtrl($scope, $http, $interval, $location, topologyServ
+        const remove_hierarchy = ['Container', 'ContainerGroup', 'ContainerReplicator', 'ContainerService',
+                                  'ContainerRoute', 'Host', 'Vm', 'ContainerNode', 'ContainerManager'];
+        // Cant use Object.values...
+        var tmp_list = Object.keys($scope.items).map(function(key) { return $scope.items[key] });
```
Could be `_.values` ;)

himdel commented on app/assets/javascripts/controllers/container_topology/container_topology_controller.js:
```diff
@@ -37,6 +37,20 @@ function ContainerTopologyCtrl($scope, $http, $interval, $location, topologyServ
+        var kind_index = 0;
+        while ((tmp_list.length > data.data.size_limit) && kind_index < remove_hierarchy.length) {
+          kind_to_hide = remove_hierarchy[kind_index];
+          tmp_list = tmp_list.filter(function(item) {
```
Not a good idea to change `tmp_list` while iterating over it..

himdel commented on app/assets/javascripts/controllers/container_topology/container_topology_controller.js:
```diff
@@ -37,6 +37,20 @@ function ContainerTopologyCtrl($scope, $http, $interval, $location, topologyServ
+        var tmp_list = Object.keys($scope.items).map(function(key) { return $scope.items[key] });
+        var kind_index = 0;
+        while ((tmp_list.length > data.data.size_limit) && kind_index < remove_hierarchy.length) {
+          kind_to_hide = remove_hierarchy[kind_index];
```
missing `var`

himdel commented on app/views/configuration/_ui_1.html.haml:
```diff
@@ -61,10 +61,11 @@
+             [_("Tile View"),    "perpage_tile",     :tile],
+             [_("List View"),    "perpage_list",     :list],
+             [_("Reports"),      "perpage_reports",  :reports],
+             [_("Topolgy View"), "perpage_topology", :topology]].each do |item_per_page|
```
`Topolgy` != `Topology` ;)

zeari commented:
@miq-bot remove_label wip

@simon3z  Aside from the comment by @serenamarie125 This should be ready.

serenamarie125 commented:
@zeari i would suggest something more like this, where it's in a box on it's own.
Here i have a switch which allows them to say if they want the auto filtering first ... is this something you could do as well ?  If set to off, then the drop down would be disabled

![screen shot 2017-01-19 at 10 57 12 am](https://cloud.githubusercontent.com/assets/12504403/22114022/1f49cf3e-de36-11e6-98e5-50eacc478332.png)

himdel commented on app/assets/javascripts/controllers/container_topology/container_topology_controller.js:
```diff
@@ -37,6 +37,20 @@ function ContainerTopologyCtrl($scope, $http, $interval, $location, topologyServ
       if (currentSelectedKinds && (Object.keys(currentSelectedKinds).length != Object.keys($scope.kinds).length)) {
         $scope.kinds = currentSelectedKinds;
+      } else {
+        remove_hierarchy = ['Container', 'ContainerGroup', 'ContainerReplicator', 'ContainerService',
```
I guess this should be

```
var remove_hierarchy = [
  'Container',
  'ContainerGroup',
  'ContainerManager',
  'ContainerNode',
  'ContainerReplicator',
  'ContainerRoute',
  'ContainerService',
  'Host',
  'Vm',
];
```

himdel replied:
(because missing var, improbable formatting, and easier to see if its one entry per line, sorted alphabetically)

simon3z commented:
> @zeari i would suggest something more like this, where it's in a box on it's own.
> Here i have a switch which allows them to say if they want the auto filtering first ... is this something you could do as well ? If set to off, then the drop down would be disabled

@serenamarie125 what about having an "Unlimited" item in the list to select from? (e.g. 20, 50, 100, unlimited)

serenamarie125 commented:
probably easier to implement @simon3z, I'd be fine with that

Loicavenel commented:
@zeari Title says:  Limit number of objects on Topology View but it seems the code is about Container only. Are we addressing topology view in general here?

himdel commented:
@Loicavenel the limit itself affects all the topologies (ToplogyController on the ruby side is the base class for all topology controllers).

The additional changes to filter out some kinds of items only affects containers though.

zeari commented:
>@Loicavenel the limit itself affects all the topologies (ToplogyController on the ruby side is the base class for all topology controllers).

@himdel @Loicavenel Not quite. Similar changes need to be made on other JS topology controllers for the limit to apply there too. The owners of other providers need to decide on an order of granularity for their entities.

zeari commented:
All comments addressed.
@serenamarie125 Have a look at the new screenshots of the topology settings.

simon3z commented on app/views/configuration/_ui_1.html.haml:
```diff
@@ -77,6 +77,18 @@
+              = _("Default Maximum Items")
+            .col-md-8
+              = select_tag(_("topology_max_objects"),
+                           options_for_select([[N_("Unlimited"), 0]].concat(PPCHOICES), @edit[:new][:topology][:max_objects]),
```
@zeari it seems you could create a `PPCHOICES_UNLIMITED` (or `PPCHOICES3`?) [here](https://github.com/ManageIQ/manageiq-ui-classic/blob/3679925/app/helpers/ui_constants.rb#L58). Otherwise other engineers will hardly discover this additional value we're introducing. Also I think over 200 items is already "unlimited" for the topology.

himdel commented:
> Similar changes need to be made on other JS topology controllers for the limit to apply there too. The owners of other providers need to decide on an order of granularity for their entities.

Oh, sorry, I misunderstood that change..

So, are you planning to do a similar change for the other topologies or is that a TODO item for us? :) (Because, merging this means we suddenly have a bug where a global topology setting does not affect 4 of the 5 topology views :).)

simon3z commented:
> So, are you planning to do a similar change for the other topologies or is that a TODO item for us? :) (Because, merging this means we suddenly have a bug where a global topology setting does not affect 4 of the 5 topology views :).)

@dclarizio @blomquisg @chessbyte how do you handle these UI changes that affect many providers? (Understanding of entities of those providers is needed... so I don't know if we can cover this effectively).

serenamarie125 commented:
@himdel i wonder if this should be changed to just have a setting for the Container Topology then?  That way, it would be an RFE to get this into Network, etc ?   I think based on environments, it would make sense to have it be a separate setting per area.  

@zeari looks good, but we may want to consider having it a Container Topology setting

himdel commented:
> i wonder if this should be changed to just have a setting for the Container Topology then?

That'd make sense to me for now :).

simon3z commented:
@serenamarie125 @himdel so it's just about renaming the name of the setting, right?

zeari commented:
1) Changed to containers specifically. - take another look at the screenshot
2) @himdel I made the change as generic as possible, i also moved some functionality into topologyService. As soon as we like this solution we can open an issue for other topologies.

simon3z commented:
@zeari what's the default? Unlimited? I think that the default should be 50, the user is still allowed to add more items by enabling them in the top bar, right?

himdel commented on spec/javascripts/fixtures/json/container_topology_response.json:
```diff
@@ -349,5 +349,8 @@
-      "OpenshiftEnterprise":{"type":"image","icon":"...openshift_enterprise..."}}}
+      "OpenshiftEnterprise":{"type":"image","icon":"...openshift_enterprise..."}
+    },
+    "size_limit": 100
```
I think this is no longer the right data, hence the spec failure ;)

himdel commented:
Thanks @zeari that looks really nice, should make it much easier to add to another topologies :).

I'd merge, except for the spec failure :)

simon3z commented:
> I'd merge, except for the spec failure :)

@himdel please note that above there's still a comment on the default value.

himdel commented:
Ah, yeah, sorry about that, the default value is up to you guys, if that's something for a longer discussion, changing it would be trivial so .. I don't really consider it a blocker :) (But the current detail would seem to be Unlimited.)

himdel commented:
@zeari One more thing I notice .. if you enter the Settings screen with a clean (new) user, you'll get an exception...

`undefined method `[]' for nil:NilClass [configuration/index]`

Pretty sure it's that `@edit[:new][:perpage][item_per_page[2]]` in the `ui_1` view.

zeari commented:
> @zeari One more thing I notice .. if you enter the Settings screen with a clean (new) user, you'll get an exception...
> 
> undefined method '[]' for nil:NilClass [configuration/index]
> 
> Pretty sure it's that @edit[:new][:perpage][item_per_page[2]] in the ui_1 view.
> 

I made a new user but im not getting this exception. `ui_constants` is reloaded when you reset the server so it might be that. topology setting was under perpage but isnt anymore so a restart might kick it off.

> (Note I also have to change entity.default_authentications.status.capitalize to ....try(:capitalize) (ContainerTopologyService#entity_status), but that's not a problem for this PR.)

How did you add your containers provider? this shouldnt happen anymore since the ui doesnt let you add a containers provider that doesnt have a bearer auth.

zeari commented:
> @himdel please note that above there's still a comment on the default value.

@simon3z Changed default value to `unlimited`.

himdel commented:
> @himdel I made a new user but im not getting this exception. ui_constants is reloaded when you reset the server so it might be that. topology setting was under perpage but isnt anymore so a restart might kick it off.

You were right, sorry about that, it was indeed that restart thing :).


> How did you add your containers provider? this shouldnt happen anymore since the ui doesnt let you add a containers provider that doesnt have a bearer auth.


Well, I'm not seeing any migrations that would fix providers that were added before it couldn't happen ;). But yeah, my db is not new.

himdel commented:
Ok, this is working and looks good, merging.

abonas commented:
@miq-bot add_label topology
